### PR TITLE
[FLINK-28993][table-planner] Fix adjusting join cost for dpp query pattern error

### PR DESCRIPTION
**Pull request by swuferhong**
## What is the purpose of the change

Now, adjusting join cost for dpp query pattern[FLINK-28821] have two errors:

1. For these table which without table stats, we haven't  multiply by  the `dynamicPartitionPruningFactor`.

2. For method `#DynamicPartitionPruningUtils$visitFactSide`, we have not judge whether `candiateFields` contain partition keys.

## Brief change log

- For these table which without table stats, we also multiply by the `dynamicPartitionPruningFactor`.
- Adding judgment for `candiateFields`.

## Verifying this change

- Need to add.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature?  no
  - If yes, how is the feature documented?  no

### CODE REVIEW

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java [hunk 1]
```diff
@@ -171,7 +171,20 @@ private static void visitFactSide(
...
                             .collect(Collectors.toList());
-            factSideFactors.isSuitableFactScanSource = !candidateFields.isEmpty();
+            if (candidateFields.isEmpty()) {
+                factSideFactors.isSuitableFactScanSource = false;
+                return;
+            }
+
+            List<String> suitableFields = new ArrayList<>();
+
+            for (String candidateField : candidateFields) {
```
Comment (godfreyhe) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java:181]: please add some comments to explain why we need this change

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/rules/physical/batch/DynamicPartitionPruningRuleTest.java [hunk 2]
```diff
@@ -410,21 +419,35 @@ public void testComplexDimSideWithAggInDimSide() {
...
 
     @Test
-    public void testDPPFactorToReorderTable() {
+    public void testDPPFactorToReorderTableWithoutStats() {
         // While there are several joins, and fact table not adjacent to dim table directly. dynamic
         // partition pruning factor will try best to reorder join relations to make fact table
         // adjacent to dim table.
-        String ddl1 =
-                "CREATE TABLE test_database.sales (\n"
+        String ddl2 =
```
Comment (godfreyhe) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/rules/physical/batch/DynamicPartitionPruningRuleTest.java:426]: nit: ddl2 -> ddl ?

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/rules/physical/batch/DynamicPartitionPruningRuleTest.java [hunk 3]
```diff
@@ -479,10 +543,20 @@ public void testDPPFactorWithDimSideJoinKeyChanged() {
 
         String query =
                 "Select * from fact_part join item on fact_part.id = item.id"
-                        + " join sales on fact_part.id = sales.id"
                         + " join (select dim_date_sk + 1 as dim_date_sk, price from dim) dim1"
                         + " on fact_part.fact_date_sk = dim1.dim_date_sk"
                         + " where dim1.price < 500 and dim1.price > 300";
         util.verifyRelPlan(query);
     }
+
```
Comment (godfreyhe) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/rules/physical/batch/DynamicPartitionPruningRuleTest.java:551]: please add a test which join keys are not partition key but are directly reference the face source fields

**Review by godfreyhe — COMMENTED** [review]
Thanks for the contribution, LGTM overall except a few minor comments

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java (hunk 1 above)
Comment (swuferhong) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java:181]: Done!

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/rules/physical/batch/DynamicPartitionPruningRuleTest.java (hunk 2 above)
Comment (swuferhong) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/rules/physical/batch/DynamicPartitionPruningRuleTest.java:426]: > nit: ddl2 -> ddl ?

Done!

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/rules/physical/batch/DynamicPartitionPruningRuleTest.java (hunk 3 above)
Comment (swuferhong) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/rules/physical/batch/DynamicPartitionPruningRuleTest.java:551]: > please add a test which join keys are not partition key but are directly reference the face source fields

Done!

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java (hunk 1 above)
Comment (godfreyhe) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java:183]: That is not the real reason. The reason is many unexpected query pattern will be treated as dpp if the candidateFields are not partition keys.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsDynamicFiltering.java [hunk 4]
```diff
@@ -60,9 +60,16 @@
 public interface SupportsDynamicFiltering {
 
     /**
-     * Applies the candidate filter fields into the table source, and return the accepted fields.
-     * The data corresponding the filter fields will be provided in runtime, which can be used to
-     * filter the partitions or the input data.
+     * Return the filter fields this partition table source supported. This method is used in
+     * planner, which can be used to judge whether the source table is a suitable partition table
+     * for filtering partitions.
```
Comment (godfreyhe) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsDynamicFiltering.java:65]: /**
     * Returns the filter fields this partition table source supported. This method can tell the
     * planner which fields can be used as dynamic filtering fields, the planner will pick some
     * fields from the returned fields based on the query, and create dynamic filtering operator.
     */

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsDynamicFiltering.java [hunk 5]
```diff
@@ -60,9 +60,16 @@
...
      */
-    List<String> applyDynamicFiltering(List<String> candidateFilterFields);
+    List<String> listAcceptedFilterFields();
+
+    /**
+     * Applies the candidate filter fields into the table source. The data corresponding the filter
+     * fields will be provided in runtime, which can be used to filter the partitions or the input
+     * data.
+     */
```
Comment (godfreyhe) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsDynamicFiltering.java:73]: /**
     * Applies the candidate filter fields into the table source. The data corresponding the filter
     * fields will be provided in runtime, which can be used to filter the partitions or the input
     * data.
     *
     * `<p>`NOTE: the candidate filter fields are always from the result of {@link
     * #listAcceptedFilterFields()}.
     */

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveTableSource.java [hunk 6]
```diff
@@ -252,48 +250,41 @@ public void applyPartitions(List<Map<String, String>> remainingPartitions) {
...
-                    dynamicFilterPartitionKeys.add(field);
-                }
-            }
-            if (dynamicFilterPartitionKeys.isEmpty()) {
-                LOG.warn(
-                        "No dynamic filter field is accepted,"
-                                + " only partition fields can use for dynamic filtering.");
-            }
+    public List<String> listAcceptedFilterFields() {
+        return catalogTable.getPartitionKeys();
```
Comment (godfreyhe) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveTableSource.java:254]: only return the partition keys with the supported types

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveTableSource.java [hunk 7]
```diff
@@ -252,48 +250,41 @@ public void applyPartitions(List<Map<String, String>> remainingPartitions) {
...
+            if (HiveSourceDynamicFileEnumerator.SUPPORTED_TYPES.contains(
+                    catalogTable
+                            .getSchema()
+                            .getFieldDataType(field)
+                            .map(DataType::getLogicalType)
+                            .map(LogicalType::getTypeRoot)
+                            .orElse(null))) {
+                dynamicFilterPartitionKeys.add(field);
+            }
         }
```
Comment (godfreyhe) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveTableSource.java:271]: move them into `listAcceptedFilterFields` method

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveTableSource.java [hunk 8]
```diff
@@ -252,48 +250,41 @@ public void applyPartitions(List<Map<String, String>> remainingPartitions) {
...
-            checkState(
-                    this.dynamicFilterPartitionKeys == null
-                            || this.dynamicFilterPartitionKeys.equals(dynamicFilterPartitionKeys),
-                    "Dynamic filtering is applied twice but with different keys: %s != %s",
-                    this.dynamicFilterPartitionKeys,
-                    dynamicFilterPartitionKeys);
-
-            this.dynamicFilterPartitionKeys = dynamicFilterPartitionKeys;
-            return dynamicFilterPartitionKeys;
-        } else {
```
Comment (godfreyhe) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveTableSource.java:293]: if it's not a partition table, we can throw exception directly

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java [hunk 9]
```diff
@@ -171,7 +171,29 @@ private static void visitFactSide(
...
                             .collect(Collectors.toList());
-            factSideFactors.isSuitableFactScanSource = !candidateFields.isEmpty();
+            if (candidateFields.isEmpty()) {
+                factSideFactors.isSuitableFactScanSource = false;
+                return;
+            }
+
+            List<String> acceptedFilterFields =
+                    ((SupportsDynamicFiltering) tableSource).listAcceptedFilterFields();
+            if (acceptedFilterFields == null) {
```
Comment (godfreyhe) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java:181]: acceptedFilterFields == null || acceptedFilterFields.isEmpty()

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/physical/batch/DynamicPartitionPruningRule.java [hunk 10]
```diff
@@ -138,30 +137,26 @@ private static List<Integer> getAcceptedFieldIndices(
...
-                        String.format(
-                                "Field: %s does not exist in the given fields: %s, "
-                                        + "please verify the applyDynamicFiltering method in connector: %s",
-                                field, candidateFields, tableSource.asSummaryString()));
+        List<String> acceptedFilterFields = new ArrayList<>();
+        for (String candidateField : candidateFields) {
+            if (acceptedFilterFieldsInSource.contains(candidateField)) {
+                acceptedFilterFields.add(candidateField);
             }
         }
```
Comment (godfreyhe) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/physical/batch/DynamicPartitionPruningRule.java:148]: extract a common method which can be reused by DynamicPartitionPruningRule

**Review by godfreyhe — COMMENTED** [review]
Thanks for the update, I left some comments

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsDynamicFiltering.java (hunk 4 above)
Comment (swuferhong) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsDynamicFiltering.java:65]: > /** * Returns the filter fields this partition table source supported. This method can tell the * planner which fields can be used as dynamic filtering fields, the planner will pick some * fields from the returned fields based on the query, and create dynamic filtering operator. */

Done!

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsDynamicFiltering.java (hunk 5 above)
Comment (swuferhong) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsDynamicFiltering.java:73]: > /** * Applies the candidate filter fields into the table source. The data corresponding the filter * fields will be provided in runtime, which can be used to filter the partitions or the input * data. * * `<p>`NOTE: the candidate filter fields are always from the result of {@link * #listAcceptedFilterFields()}. */

Done!

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveTableSource.java (hunk 6 above)
Comment (swuferhong) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveTableSource.java:254]: > only return the partition keys with the supported types

Done!

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveTableSource.java (hunk 8 above)
Comment (swuferhong) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveTableSource.java:293]: > if it's not a partition table, we can throw exception directly

Done!

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java (hunk 9 above)
Comment (swuferhong) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/DynamicPartitionPruningUtils.java:181]: > acceptedFilterFields.isEmpty()

Done!

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/physical/batch/DynamicPartitionPruningRule.java (hunk 10 above)
Comment (swuferhong) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/physical/batch/DynamicPartitionPruningRule.java:148]: > extract a common method which can be reused by DynamicPartitionPruningRule

Done!

**Comment by swuferhong** [comment]
@flinkbot run azure

**Review by godfreyhe — APPROVED** [review]
Thanks for the update, LGTM

**Comment by godfreyhe** [comment]
@flinkbot run azure

**Comment by swuferhong** [comment]
@flinkbot run azure

**Comment by swuferhong** [comment]
@flinkbot run azure

**Comment by godfreyhe** [comment]
@flinkbot run azure

**Comment by swuferhong** [comment]
@flinkbot run azure

**Comment by godfreyhe** [comment]
@flinkbot run azure

**Pull request closed without merging**